[chore] add container for admin settings form

Kharonus opened the pull request:
# Ticket
Reference implementation for [OP#61755](https://community.openproject.org/wp/61755)

# What are you trying to accomplish?
- harmonize visuals of admin setting pages

# What approach did you choose and why?
- admin setting forms can use as base component `AdminSettingsFormComponent` and will benefit from the general approved form container for admin setting pages
- applied for wp types general settings and wp types subject generation

myabc commented on app/components/admin_settings_form_component.sass:
```diff
@@ -0,0 +1,3 @@
+.op-admin-settings-form-wrapper
+  max-width: 480px
+  margin: 0 auto 0 0
```
@Kharonus nitpick - can you add a new line? (I think that's what our `.editorconfig` specifies)

Kharonus replied:
It has an empty newline, otherweise github would show this fancy red circle icon.

myabc commented on app/components/admin_settings_form_component.sass:
```diff
@@ -0,0 +1,3 @@
+.op-admin-settings-form-wrapper
+  max-width: 480px
```
How did you arrive at this number? 

I know it's hard to avoid "magic numbers" in CSS, but in this case defining a variable *might* help document the intent?

btw, I see in `_forms.sass` we are setting a max width of 500px for the enterprise banner:

```sass
  // follows new board page columns width (which uses -wide class (500px) for its tiles)
  .boards-enterprise-banner-wide
    max-width: calc(2*500px + 1rem)

  @media screen and (max-width: $breakpoint-md)
    .boards-enterprise-banner-wide
      max-width: 500px
```

HDinger replied:
Yeah, we probably should discuss that number. It feels too small for me..

Kharonus replied:
I just took it from figma. For me, the technical approach is more important, then the real number. The number can be changed after we discuss it with UI/UX.

HDinger replied:
we decided in todays call to go with 680 for now. That can still be changed later on.

Kharonus commented on app/components/admin_settings_form_component.sass:
```diff
@@ -0,0 +1,3 @@
+.op-admin-settings-form-wrapper
+  max-width: 480px
```
❓ is there no variable for this? 🤷🏽

myabc reviewed (commented):
@Kharonus Why not go for a simpler approach? e.g. applying a class in the templates `app/components/work_packages/types/settings_component.html.erb` and `app/components/work_packages/types/subject_configuration_component.html.erb`

```erb
<%=
  primer_form_with(**form_options, class: "op-admin-settings-form") do |f|
    render(WorkPackages::Types::SettingsForm.new(f))
  end
%>
```

(Of course you could still pull up the common behaviour into a parent `AdminSettingsFormComponent` and then define `form_options`).

We could also consider defining a wrapper component - something like this:

```ruby
class AdminSettingsFormWrapper < ApplicationComponent
  def initialize(model, **system_arguments)
    @system_arguments = system_arguments
    @system_arguments[:tag] = :div
    @system_arguments[:classes] = class_names(
      @system_arguments[:classes],
      "op-admin-settings-form-wrapper"
    )
    super
  end

  def call
    BaseComponent.new(**@component_arguments).render_in(self) { content }
  end
end
```

and then manually wrapping the Primer forms in applicable templates:

```erb
<%=
  admin_settings_form_wrapper do
    primer_form_with(**form_options, class: "op-admin-settings-form") do |f|
    render(WorkPackages::Types::SettingsForm.new(f))
    end
  end
%>
```

HDinger commented:
> @Kharonus Why not go for a simpler approach? e.g. applying a class in the templates `app/components/work_packages/types/settings_component.html.erb` and `app/components/work_packages/types/subject_configuration_component.html.erb`
> 
> ```
> <%=
>   primer_form_with(**form_options, class: "op-admin-settings-form") do |f|
>     render(WorkPackages::Types::SettingsForm.new(f))
>   end
> %>
> ```

@Kharonus and I discussed that approach as well, but I think this is too "weak".. Building a whole pattern only on the class being is not ideal because of three reasons:
* A class is is not very visible. It is not clear that this class is essential for all setting pages to look the same. They tend to be forgotten or duplicated later on because no ones about its existence. Components and wrappers have a bigger visibility.
* Classes also tend to be easier changed/deleted than components or wrappers.
* In case, we want to modify further stuff in the settings form pages, we can do that easily with the component.



> (Of course you could still pull up the common behaviour into a parent `AdminSettingsFormComponent` and then define `form_options`).
> 
> We could also consider defining a wrapper component - something like this:
> 
> ```ruby
> class AdminSettingsFormWrapper < ApplicationComponent
>   def initialize(model, **system_arguments)
>     @system_arguments = system_arguments
>     @system_arguments[:tag] = :div
>     @system_arguments[:classes] = class_names(
>       @system_arguments[:classes],
>       "op-admin-settings-form-wrapper"
>     )
>     super
>   end
> 
>   def call
>     BaseComponent.new(**@component_arguments).render_in(self) { content }
>   end
> end
> ```
> 
> and then manually wrapping the Primer forms in applicable templates:
> 
> ```
> <%=
>   admin_settings_form_wrapper do
>     primer_form_with(**form_options, class: "op-admin-settings-form") do |f|
>     render(WorkPackages::Types::SettingsForm.new(f))
>     end
>   end
> %>
> ```

I could live with that approach. Wdyt @Kharonus ?

HDinger reviewed (commented):
I like the way this going 👍 I think we have to change the value itself. Maybe we can also discuss whether we want to offer multiple size options, but that can be done separately.

Kharonus commented:
I'm not really sure, I understand everything from Alex' suggestion, but I understand the consequence:

instead of defining the overall layout in the component by inheriting from another base class, we would wrap the content in the template. Honestly, I do not have any preference. What are the pros and cons to either?

Once we decided on the correct technical approach, I can amend the lookbook docs, too.

Kharonus commented:
So, @HDinger and @myabc, can we get a decision here? Wanna have some vote somewhere? :D

HDinger commented:
I don't have a strong opinion on this. 

I see two benefits in the wrapper:

* As of know, the whole component/wrapper creates only an additional HTML element. So putting it in the HMTL seems very logical. Of course, that might change as soon the wrapper does more than that.
* You don't need a seperate component which inherits but can use the element directly in any template. I don't know whether that case is really useful or whether we will always have components around 🤷‍♀️ 

Both arguments are somehow weak, I admit. But when they are no counter arguments, I'd vote to go with the wrapper.

Kharonus committed:
[chore] add container for admin settings form

- admin setting forms can use as base component
  `AdminSettingsFormComponent` and will benefit from the general
  approved form container for admin setting pages

Kharonus committed:
[chore] added documentation changes

Kharonus committed:
[chore] updated settings pages of lookbook

Kharonus committed:
[chore] added 680px as base value for admin settings form width

Kharonus commented:
Added @psatyal and a pull preview, so that we can review the doc changes and check the looks of the example forms (wp types).

psatyal commented:
Hi @Kharonus. Yes, I think the wrapper makes things a lot cleaner, for example in the Work package types page ('Settings' tab). I noticed that the forms to editing a work package status or the 'Progress tracking' page don't have the wrapper; should we create a separate PR/feature to adding the wrapper to other pages, or can we do it in this PR?

Thanks for the feature btw! Makes things look a lot nicer.

Kharonus commented:
I'd like to have another feature for adding it to existent pages. Then we can have a clear list for QA to check all pages for the style.

this PR is not tied to a feature wp, as it was meant as a "chore", a preparation work.

HDinger commented on lookbook/docs/patterns/17-settings-pages.md.erb:
```diff
@@ -1 +1,29 @@
+such a wrapping container.
+
+Any view component rendering a form should inherit from `AdminSettingsFormComponent`. The template stays unchanged. This
+way, the rendered form is wrapped into a container with `480px` max-width.
```
```suggestion
way, the rendered form is wrapped into a container with `680px` max-width.
```

psatyal replied:
FYI: Made this change my (later) commit.

HDinger reviewed (approved):
Good from my side, just one minor remark. 👍

psatyal committed:
Minor changes to phrasing

psatyal committed:
Minor changes in phrasing and correction: 680px

psatyal reviewed (approved):
Hi @Kharonus. Made a few minor changes, looks good to me. Thanks!
